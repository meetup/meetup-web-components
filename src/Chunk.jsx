import React from 'react';
import cx from 'classnames';

/**
 * @module Chunk
 */
class Chunk extends React.Component {
	render() {
		const {
			children,
			className,
			...other
		} = this.props;

		const classNames = cx(
			'chunk',
			className
		);

		return (
			<div
				className={classNames}
				{...other}>
					{children}
			</div>
		);
	}
}

Chunk.propTypes = {
};

export default Chunk;
