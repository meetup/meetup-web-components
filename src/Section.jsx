import React from 'react';
import cx from 'classnames';

/**
 * @module Section
 */
class Section extends React.Component {
	render() {
		const {
			children,
			className,
			...other
		} = this.props;

		const classNames = cx(
			'section',
			className
		);

		return (
			<div
				className={classNames}
				{...other}>
					{children}
			</div>
		);
	}
}

Section.propTypes = {

};

export default Section;
