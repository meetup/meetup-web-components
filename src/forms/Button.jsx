import React from 'react';
import cx from 'classnames';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

export const BUTTON_CLASS = 'button';

/**
 * SQ2 Button component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_buttons.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#buttons}
 * @module Button
 */
class Button extends React.Component {

	render() {
		const {
			children,
			className,
			onClick,
			reset,
			contrast,
			fullWidth,
			icon,
			primary,
			right,
			small,
			...other
		} = this.props;

		const classNames = cx(
			BUTTON_CLASS,
			{
				'button--contrast': contrast,
				'button--fullWidth': fullWidth,
				'button--primary': primary,
				'button--small': small,
				'button--reset': reset,
			},
			className
		);
		const opts = right ? { rowReverse: 'all' } : {};

		const iconChildren = (
			<Flex className='button--icon-wrapper' {...opts}>
				{icon &&
					<FlexItem shrink className='button--icon valign--middle'>
						{icon}
					</FlexItem>
				}
				{children &&
					<FlexItem className='button--label valign--middle align--center atMedium_align--left'>
						{children}
					</FlexItem>
				}
			</Flex>
		);

		return (
			<button
				className={classNames}
				onClick={onClick}
				role='button'
				{...other}
			>
				{ icon ? iconChildren : children }
			</button>
		);
	}
}

Button.defaultProps = {
	right: false
};

Button.propTypes = {
	reset: React.PropTypes.bool,
	contrast: React.PropTypes.bool,
	fullWidth: React.PropTypes.bool,
	primary: React.PropTypes.bool,
	small: React.PropTypes.bool,
	icon: React.PropTypes.any,
	right: React.PropTypes.bool,
};
export default Button;
