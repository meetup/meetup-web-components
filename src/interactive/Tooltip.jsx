import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import Button from '../forms/Button';
import Icon from '../media/Icon';
import FloatingPosition from '../utils/components/FloatingPosition';

/**
 * @module Tooltip
 */
class Tooltip extends React.PureComponent {
	constructor(props) {
		super(props);

		this.onBlur = this.onBlur.bind(this);
		this.closeContent = this.closeContent.bind(this);
		this.openContent = this.openContent.bind(this);
		this.handleClose = this.handleClose.bind(this);

		this.state = {
			isActive: props.isActive || false,
		};
	}

	static getDerivedStateFromProps(nextProps, state) {
		if (nextProps.manualToggle && state.isActive !== nextProps.isActive) {
			return { isActive: nextProps.isActive };
		}
		return null;
	}

	closeContent(e) {
		if (this.props.manualToggle) {
			if (this.props.onClick) {
				this.props.onClick(e);
				return;
			}
			this.props.onClose && this.props.onClose(e);
			return;
		}

		this.setState(() => ({ isActive: false }));

		if (this.props.onBlur) {
			this.props.onBlur(e);
		}
		if (this.props.onMouseLeave) {
			this.props.onMouseLeave(e);
		}
	}

	openContent(e) {
		this.setState(() => ({ isActive: true }));

		if (this.props.onFocus) {
			this.props.onFocus(e);
		}
		if (this.props.onMouseEnter) {
			this.props.onMouseEnter(e);
		}
	}

	onBlur(e) {
		if (!this.contentRef) return;

		setTimeout(() => {
			if (!this.contentRef.contains(document.activeElement)) {
				this.closeContent(e);
			}
		}, 0);
	}

	handleClose(e) {
		if (e.key === 'Escape') {
			this.props.onClose && this.props.onClose(e);
			this.setState(() => ({ isActive: false }));
		}
	}

	render() {
		const {
			className,
			trigger,
			content,
			align,
			offset,
			maxWidth,
			minWidth,
			noPortal,
			id,
			direction,
			manualToggle,
			withClose,
			...other
		} = this.props;

		const { isActive } = this.state;

		// Do not pass along to children
		delete other.isActive;
		delete other.onClick;

		const classNames = {
			dropdown: cx(className, 'popup', {
				'popup--noPortal': noPortal,
			}),
		};

		const getTrigger = () => {
			return this.triggerRef;
		};

		const getContent = () => {
			return this.contentRef;
		};

		return (
			<div
				className={classNames.dropdown}
				onMouseLeave={manualToggle ? undefined : this.closeContent}
				{...other}
			>
				<div
					ref={el => (this.triggerRef = el)}
					className="popup-trigger"
					onFocus={manualToggle ? undefined : this.openContent}
					onBlur={manualToggle ? undefined : this.onBlur}
					onMouseEnter={manualToggle ? undefined : this.openContent}
					aria-labelledby={isActive ? id : undefined}
					role={isActive ? 'button' : undefined}
				>
					{trigger}
				</div>
				{isActive && (
					<FloatingPosition
						getTrigger={getTrigger}
						getContent={getContent}
						noPortal={noPortal}
						align={align}
						offset={offset}
						direction={direction}
						handleClose={this.handleClose}
					>
						{({ top, left, align, boundedMaxWidth }) => (
							<div
								ref={el => (this.contentRef = el)}
								className={cx('popup-content tooltip-content', {
									'popup-content--right': align === 'right',
									'popup-content--left': align === 'left',
									'popup-content--center': align === 'center',
									'popup-content--top': direction === 'top',
									'display--none': !isActive,
									'display--block': isActive,
								})}
								aria-hidden={!isActive}
								id={id}
								role="tooltip"
								style={{
									left,
									top,
									minWidth,
									maxWidth: maxWidth || boundedMaxWidth,
								}}
							>
								<div
									className={cx(
										'popup-bubble tooltip-bubble inverted',
										{
											'popup-bubble--active': isActive,
											'popup-bubble--right': align === 'right',
											'popup-bubble--left': align === 'left',
											'popup-bubble--center': align === 'center',
											'popup-bubble--top': direction === 'top',
										}
									)}
								>
									{withClose && (
										<Button
											className="tooltip-closeBtn"
											onClick={this.closeContent}
											icon={
												<Icon
													shape="cross"
													className="text--secondary"
												/>
											}
											reset
											aria-label="Close"
											role="button"
										/>
									)}
									{content}
								</div>
							</div>
						)}
					</FloatingPosition>
				)}
			</div>
		);
	}
}

Tooltip.defaultProps = {
	direction: 'bottom',
	minWidth: '0px',
	noPortal: false,
};

Tooltip.propTypes = {
	/** The unique identifier for the Tooltip */
	id: PropTypes.string.isRequired,

	/** The element that opens the tooltip when hovered or focused */
	trigger: PropTypes.element.isRequired,

	/** The content that's rendered inside the tooltip */
	content: PropTypes.element,

	/** The horizontal alignment of the tooltip content bubble to the dropdown trigger (defaults to right) */
	align: PropTypes.oneOf(['left', 'right', 'center']),

	/** How many additional pixels to push the tooltip content bubble */
	offset: PropTypes.shape({
		left: PropTypes.number,
		top: PropTypes.number,
	}),

	/** Which side of the tooltip trigger the dropdown content bubble renders on */
	direction: PropTypes.oneOf(['top', 'bottom']).isRequired,

	/** Class names to add to the wrapper of the tooltip trigger and content */
	className: PropTypes.string,

	/** Whether the tooltip content is being shown */
	isActive: PropTypes.bool,

	/** Whether the Tooltip is a controlled component, relying on `isActive` prop for toggling */
	manualToggle: PropTypes.bool,

	/** The largest width the tooltip content can be */
	maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

	/** The smallest width the tooltip content can be */
	minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

	/** Whether to render the dropdown content directly in the component instead of bulling it out and attaching to the document root */
	noPortal: PropTypes.bool,
	/** Required function to close the dropdown when working with a manually toggled tooltip */
	onClose: (props, propName, componentName) => {
		if (
			props.manualToggle === true &&
			(props[propName] === undefined || typeof props[propName] !== 'function')
		) {
			return new Error(
				`The ${propName} function is required when manualToggle is true in this component!`
			);
		}
	},
};

export default Tooltip;
