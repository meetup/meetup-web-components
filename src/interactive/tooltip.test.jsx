import React from 'react';
import { shallow, mount } from 'enzyme';
import Section from '../layout/Section';
import Chunk from '../layout/Chunk';
import Button from '../forms/Button';
import { Icon as SwarmIcon } from '@meetup/swarm-components';

import Tooltip from './Tooltip';

const dropdownContent = (
	<Section noSeparator>
		<Chunk>
			<h2 className="text--big text--bold">Dropdown content</h2>
		</Chunk>
	</Section>
);
const dropdownTrigger = <Button small>Open</Button>;
const dropdownId = 'testDropdown';

describe('Tooltip', () => {
	const dropdownJSX = (
		<Tooltip
			align="right"
			id={dropdownId}
			trigger={dropdownTrigger}
			content={dropdownContent}
		/>
	);
	const wrapper = shallow(dropdownJSX);

	it('should hide dropdown content by default', () => {
		expect(wrapper.find('.popup-content').length).toBeFalsy();
	});

	it('should call onMouseEnter and onMouseLeave functions if passed as props', () => {
		const onMouseEnter = jest.fn();
		const onMouseLeave = jest.fn();

		const dropdownMouseEvents = shallow(
			<Tooltip
				align="right"
				id={dropdownId}
				trigger={dropdownTrigger}
				content={dropdownContent}
				onMouseEnter={onMouseEnter}
				onMouseLeave={onMouseLeave}
			/>
		);
		const trigger = dropdownMouseEvents.find('.popup-trigger').first();
		const tooltipArea = dropdownMouseEvents.find('.popup').first();

		expect(onMouseEnter).not.toHaveBeenCalled();
		expect(onMouseLeave).not.toHaveBeenCalled();
		trigger.simulate('mouseEnter');
		expect(onMouseEnter).toHaveBeenCalled();
		tooltipArea.simulate('mouseLeave');
		expect(onMouseLeave).toHaveBeenCalled();
	});

	it('should call onFocus and onBlur functions if passed as props', () => {
		const onFocus = jest.fn();
		const onBlur = jest.fn();

		const dropdownFocusEvents = shallow(
			<Tooltip
				align="right"
				id={dropdownId}
				trigger={dropdownTrigger}
				content={dropdownContent}
				onFocus={onFocus}
				onBlur={onBlur}
			/>
		);
		const trigger = dropdownFocusEvents.find('.popup-trigger').first();
		const tooltipArea = dropdownFocusEvents.find('.popup').first();

		expect(onFocus).not.toHaveBeenCalled();
		expect(onBlur).not.toHaveBeenCalled();
		trigger.simulate('focus');
		expect(onFocus).toHaveBeenCalled();
		tooltipArea.simulate('blur');
		setTimeout(() => {
			expect(onBlur).toHaveBeenCalled();
		}, 10);
	});

	describe('aligned tooltip', () => {
		const rightDropdown = (
			<Tooltip
				align="right"
				direction="top"
				id={dropdownId}
				trigger={dropdownTrigger}
				content={dropdownContent}
			/>
		);
		const rightDropdownWrapper = mount(rightDropdown);
		const trigger = rightDropdownWrapper.find('.popup-trigger');

		it('matches snapshot', () => {
			expect(rightDropdown).toMatchSnapshot();
		});

		it('applies correct alignment className to tooltip content', () => {
			trigger.simulate('focus');
			const content = rightDropdownWrapper.find('.popup-content');

			expect(content.prop('className')).toContain('popup-content--right');
		});
	});

	describe('open and close', () => {
		let closedComponent, trigger, tooltipArea;

		beforeEach(() => {
			closedComponent = shallow(dropdownJSX);
			trigger = closedComponent.find('.popup-trigger').first();
			tooltipArea = closedComponent.find('.popup').first();
		});
		afterEach(() => {
			closedComponent = null;
			trigger = null;
			tooltipArea = null;
		});

		it('should show tooltip when trigger is hovered', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('mouseEnter');
			expect(closedComponent.state('isActive')).toBeTruthy();
		});

		it('should hide tooltip when cursor leaves the tooltip area', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('mouseEnter');
			expect(closedComponent.state('isActive')).toBeTruthy();
			tooltipArea.simulate('mouseLeave');
			setTimeout(() => {
				expect(closedComponent.state('isActive')).toBeFalsy();
			}, 1);
		});

		it('should show tooltip when trigger is focused', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('focus');
			expect(closedComponent.state('isActive')).toBeTruthy();
		});

		it('should hide tooltip when trigger is blurred', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('focus');
			expect(closedComponent.state('isActive')).toBeTruthy();
			tooltipArea.simulate('blur');

			setTimeout(() => {
				expect(closedComponent.state('isActive')).toBeFalsy();
			}, 10);
		});
	});

	describe('tooltip with a close button', () => {
		const openTooltipWithCloseJSX = (
			<Tooltip
				isActive
				withClose
				align="right"
				id={dropdownId}
				trigger={dropdownTrigger}
				content={dropdownContent}
			/>
		);
		const wrapper = mount(openTooltipWithCloseJSX);
		const closeBtn = wrapper.find(SwarmIcon);

		it('should render correctly', () => {
			expect(wrapper).toMatchSnapshot();
		});

		it('should hide tooltip close button is clicked', () => {
			expect(wrapper.state('isActive')).toBeTruthy();
			closeBtn.simulate('click');
			expect(wrapper.state('isActive')).toBeFalsy();
		});
	});

	describe('manually toggled tooltip', () => {
		const manualTooltip = (
			<Tooltip
				manualToggle
				align="right"
				id={dropdownId}
				trigger={dropdownTrigger}
				content={dropdownContent}
			/>
		);
		const manualTooltipWrapper = mount(manualTooltip);
		const trigger = manualTooltipWrapper.find('.popup-trigger');

		it('does not render content on focus', () => {
			trigger.simulate('focus');
			expect(manualTooltipWrapper.find('.popup-content').length).toBe(0);
		});
	});
});
