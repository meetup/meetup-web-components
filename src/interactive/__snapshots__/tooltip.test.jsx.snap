// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip aligned tooltip matches snapshot 1`] = `
<Tooltip
  align="right"
  content={
    <Section
      noSeparator={true}
    >
      <Chunk>
        <h2
          className="text--big text--bold"
        >
          Dropdown content
        </h2>
      </Chunk>
    </Section>
  }
  direction="top"
  id="testDropdown"
  minWidth="0px"
  noPortal={false}
  trigger={
    <Button
      small={true}
    >
      Open
    </Button>
  }
/>
`;

exports[`Tooltip tooltip with a close button should render correctly 1`] = `
<Tooltip
  align="right"
  content={
    <Section
      noSeparator={true}
    >
      <Chunk>
        <h2
          className="text--big text--bold"
        >
          Dropdown content
        </h2>
      </Chunk>
    </Section>
  }
  direction="bottom"
  id="testDropdown"
  isActive={true}
  minWidth="0px"
  noPortal={false}
  trigger={
    <Button
      small={true}
    >
      Open
    </Button>
  }
  withClose={true}
>
  <div
    className="popup"
    onMouseLeave={[Function]}
  >
    <div
      aria-labelledby="testDropdown"
      className="popup-trigger"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      role="button"
    >
      <Button
        small={true}
      >
        <Button
          small={true}
        >
          <button
            data-icon="left"
            data-swarm-button="default"
            data-swarm-size="small"
            data-swarm-width="default"
            small={true}
            type="button"
          >
            <span>
              <ChildrenWithIcon>
                Open
              </ChildrenWithIcon>
            </span>
          </button>
        </Button>
      </Button>
    </div>
    <WithMatchMedia(FloatingPosition)
      align="right"
      direction="bottom"
      getContent={[Function]}
      getTrigger={[Function]}
      handleClose={[Function]}
      noPortal={false}
    >
      <FloatingPosition
        align="right"
        direction="bottom"
        getContent={[Function]}
        getTrigger={[Function]}
        handleClose={[Function]}
        media={Object {}}
        noPortal={false}
      >
        <ConditionalWrap
          condition={true}
          wrap={[Function]}
        >
          <Portal>
            <Portal
              containerInfo={
                <div>
                  <div
                    aria-hidden="false"
                    class="popup-content tooltip-content popup-content--right display--block"
                    id="testDropdown"
                    role="tooltip"
                    style="left: 0px; top: 0px; min-width: 0px; max-width: 200px;"
                  >
                    <div
                      class="popup-bubble tooltip-bubble inverted popup-bubble--active popup-bubble--right"
                    >
                      <button
                        aria-label="Close"
                        class="tooltip-closeBtn"
                        data-icon="only"
                        data-swarm-button="reset"
                        data-swarm-size="default"
                        data-swarm-width="default"
                        role="button"
                        type="button"
                      >
                        <span>
                          <span>
                            <svg
                              class="svg svg--cross svg-icon valign--middle"
                              height="20"
                              preserveAspectRatio="xMinYMin meet"
                              role="presentation"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <use
                                xlink:href="#icon-cross--small"
                              />
                            </svg>
                          </span>
                        </span>
                      </button>
                      <section
                        class="section"
                      >
                        <div
                          class="chunk"
                        >
                          <h2
                            class="text--big text--bold"
                          >
                            Dropdown content
                          </h2>
                        </div>
                      </section>
                    </div>
                  </div>
                </div>
              }
            >
              <div
                aria-hidden={false}
                className="popup-content tooltip-content popup-content--right display--block"
                id="testDropdown"
                role="tooltip"
                style={
                  Object {
                    "left": "0px",
                    "maxWidth": "200px",
                    "minWidth": "0px",
                    "top": "0px",
                  }
                }
              >
                <div
                  className="popup-bubble tooltip-bubble inverted popup-bubble--active popup-bubble--right"
                >
                  <Button
                    aria-label="Close"
                    className="tooltip-closeBtn"
                    icon={
                      <Icon
                        className="text--secondary"
                        shape="cross"
                        size="s"
                      />
                    }
                    onClick={[Function]}
                    reset={true}
                    role="button"
                  >
                    <Button
                      aria-label="Close"
                      className="tooltip-closeBtn"
                      icon={
                        <Icon
                          className="text--secondary"
                          shape="cross"
                          size="s"
                        />
                      }
                      onClick={[Function]}
                      reset={true}
                      role="button"
                    >
                      <button
                        aria-label="Close"
                        className="tooltip-closeBtn"
                        data-icon="only"
                        data-swarm-button="reset"
                        data-swarm-size="default"
                        data-swarm-width="default"
                        onClick={[Function]}
                        reset={true}
                        role="button"
                        type="button"
                      >
                        <span>
                          <ChildrenWithIcon
                            iconShape="cross"
                          >
                            <Icon
                              shape="cross"
                              size="xs"
                            >
                              <span>
                                <Icon
                                  role="presentation"
                                  shape="cross"
                                  size="xs"
                                >
                                  <svg
                                    className="svg svg--cross svg-icon valign--middle"
                                    height="20"
                                    preserveAspectRatio="xMinYMin meet"
                                    role="presentation"
                                    style={Object {}}
                                    viewBox="0 0 20 20"
                                    width="20"
                                  >
                                    <use
                                      xlinkHref="#icon-cross--small"
                                    />
                                  </svg>
                                </Icon>
                              </span>
                            </Icon>
                          </ChildrenWithIcon>
                        </span>
                      </button>
                    </Button>
                  </Button>
                  <Section
                    noSeparator={true}
                  >
                    <SectionComponent
                      noSeparator={true}
                    >
                      <section
                        className="section"
                      >
                        <Chunk>
                          <ChunkComponent>
                            <div
                              className="chunk"
                            >
                              <h2
                                className="text--big text--bold"
                              >
                                Dropdown content
                              </h2>
                            </div>
                          </ChunkComponent>
                        </Chunk>
                      </section>
                    </SectionComponent>
                  </Section>
                </div>
              </div>
            </Portal>
          </Portal>
        </ConditionalWrap>
      </FloatingPosition>
    </WithMatchMedia(FloatingPosition)>
  </div>
</Tooltip>
`;
